sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableServiceImplTest.java

```diff
@@ -126,6 +144,451 @@ public void testRead() throws IOException {
+    ByteKey secondEnd =
+        ByteKey.copyFrom(
+            ("b" + String.format("%05d", (MINI_BATCH_ROW_LIMIT - 1)))
+                .getBytes(StandardCharsets.UTF_8));
+
+    ByteKey thirdStart = ByteKey.copyFrom(("c00000").getBytes(StandardCharsets.UTF_8));
+    ByteKey thirdEnd =
+        ByteKey.copyFrom(
+            ("c" + String.format("%05d", (MINI_BATCH_ROW_LIMIT / 2 - 1)))
+                .getBytes(StandardCharsets.UTF_8));
+
+    when(mockBigtableSource.getRanges())
+        .thenReturn(
+            Arrays.asList(
+                ByteKeyRange.of(firstStart, firstEnd),
+                ByteKeyRange.of(secondStart, secondEnd),
+                ByteKeyRange.of(thirdStart, thirdEnd)));
+    when(mockBigtableDataClient.readFlatRows(any(ReadRowsRequest.class), any()))
+        .thenAnswer(mockReadRowsAnswer(rowsInRowRanges.subList(0, MINI_BATCH_ROW_LIMIT)))
+        .thenAnswer(
+            mockReadRowsAnswer(
+                rowsInRowRanges.subList(
+                    MINI_BATCH_ROW_LIMIT, MINI_BATCH_ROW_LIMIT + MINI_BATCH_ROW_LIMIT / 2)))
+        .thenAnswer(mockReadRowsAnswer(new ArrayList<FlatRow>()));
+
+    BigtableService.Reader underTest =
+        new BigtableServiceImpl.BigtableSegmentReaderImpl(mockSession, mockBigtableSource);
+
+    underTest.start();
+    verify(mockBigtableDataClient, times(1))
+        .readFlatRows(requestCaptor.capture(), any(StreamObserver.class));
+    Assert.assertEquals(3, requestCaptor.getValue().getRows().getRowRangesCount());
+    Assert.assertEquals(
+        FlatRowConverter.convert(rowsInRowRanges.get(0)), underTest.getCurrentRow());
+    for (int i = 0; i < MINI_BATCH_ROW_LIMIT; i++) {
+      underTest.advance();
+    }
+    verify(mockBigtableDataClient, times(2))
+        .readFlatRows(requestCaptor.capture(), any(StreamObserver.class));
+    for (int i = 0; i < MINI_BATCH_ROW_LIMIT; i++) {
+      underTest.advance();
+    }
+    Assert.assertEquals(
+        FlatRowConverter.convert(rowsInRowRanges.get(rowsInRowRanges.size() - 1)),
+        underTest.getCurrentRow());
+    Assert.assertEquals(1, requestCaptor.getValue().getRows().getRowRangesCount());
+    Assert.assertFalse(underTest.advance());
+
+    underTest.close();
+    verifyMetricWasSet("google.bigtable.v2.ReadRows", "ok", 2);
+  }
+
+  /**
+   * This test checks that the buffer will stop filling up once the byte limit is reached. It will
+   * cancel the ScanHandler after reached the limit. This test completes one fill and contains one
+   * Row after the first buffer has been completed. The test cheaks the current available memory in
+   * the JVM and uses a percent of it to mock the original behavior.
+   *
+   * @throws IOException
+   */
+  @Test
+  public void testReadByteLimitBuffer() throws IOException {
+    List<FlatRow> expectedFirstRangeRows = new ArrayList<>();
+    // Max amount of memory allowed in a Row (256MB)
+    byte[] largeMemory = new byte[(int) DEFAULT_ROW_SIZE];
```

Comment:
Make maxSegmentByteSize a parameter on the read constructor so that you can pass it in in tests. Then add a factory method to get it from the jvm:

```java
class Reader {
   static Reader create(BigableSession session, BigtableSource source) {
     ...
     int maxSegmentByteSize = Runtime.getRuntime().getTotalMemory() * .1;
     return new Reader(..., maxSegmentByteSize);
   }
}

```
Then in the pipeline you replace new Reader(...) with Reader.create(..).
And in your test you can instantiate it with a much lower limit new new Reader(..., 10);

This will also make your test deterministic remove all of the noise below